tools/building.py

```diff
@@ -220,13 +220,13 @@ def lld_flags_for_executable(external_symbols):
       cmd.append('--growable-table')
 
   if not settings.SIDE_MODULE:
-    # Export these two section start symbols so that we can extract the string
-    # data that they contain.
-    cmd += [
-      '-z', 'stack-size=%s' % settings.STACK_SIZE,
-      '--initial-memory=%d' % settings.INITIAL_MEMORY,
-      '--max-memory=%d' % settings.MAXIMUM_MEMORY,
-    ]
+    cmd += ['-z', 'stack-size=%s' % settings.STACK_SIZE]
+    cmd += ['--max-memory=%d' % settings.MAXIMUM_MEMORY]
```

Comment:
Perhaps this is the problem... this flag should probably not be specified in all cases.  Specifically when `ALLOW_MEMORY_GROWTH` is not specified.

Reply:
Well - if we don't specify it, wasm-ld will do so on its own, with a default of 4GB (IIRC), which would not work well with features (specifically ASan) that depend on Emscripten's `MAXIMUM_MEMORY` being accurate.

Reply:
Actually if we don't specify this the linker will great a memory with no limits on growth.

I think maybe we need a new `--no-growable-memory` flag in the linker to make this work how we want it to work.

Otherwise the result will be that all modules produces by emscripten will not become growable by default.

We could also have the linker interpret `--max-memory=0` as non-growable?

Reply:
I opened a new llvm bug for this: https://github.com/llvm/llvm-project/issues/81932

Reply:
It is certainly possible to add a new linker feature, essentially a "wasm-ld native" equivalent of `-sALLOW_MEMORY_GROWTH=0`.

However, I would step back, and ask the question: what value would that provide? Perhaps:
1) We want to avoid the runtime overhead of a high maximum memory?
   -  I am not aware of any significant costs here, however.
2) We want to remove the WASM2JS special case?
   - But WASM2JS is a legacy mode, it doesn't make sense to add linker features just for it.
3) We want to avoid the perception issue of essentially wrong values for maximum memory with `-sALLOW_MEMORY_GROWTH=0` by default.
   - This can be solved in a cheaper way by defaulting to `INITIAL_MEMORY` for `-sALLOW_MEMORY_GROWTH=0`, i. e. only enabling `INITIAL_HEAP` by default for `-sALLOW_MEMORY_GROWTH=1`.

WDYT?

P.S. I will be going offline for today...

Reply:
I think all three of your points are valid points yes.

(1)  I believe come engines are take advantage of the fact the memory is not growable (see the wasm2js issue for example).  Non-growable memories have always been the default in emscripten and I'm not sure we want to change that.   One example of how gorwable memory could effect things is that I believe engines can/do overallocate virtual address space to avoid having the move the memory when it grows, so it creates more virtual memory pressue.

(2) This is a much smaller issue but I'd like to avoid yet more special cases in emscripten if we can.

(3) Sure, if the user specifies `-sALLOW_MEMORY_GROWTH=0` (which happens to be the default) it would be very odd if the resulting module contained a growable memory.   I guess we could change the documentation for `ALLOW_MEMORY_GROWTH` to say that it defaults to true if you specify `INITIAL_HEAP`?    And then make its usage mutually exclusive with `INITIAL_HEAP`?   Overall it seems cleaner to add a new linker flag though.

Reply:
Resolved with the latest commits.